Preserve DV-specific fields for deletion vectors

adawrapub opened the pull request:
This is to fix problem reported in https://github.com/apache/iceberg/issues/13671 by preserving DV specific fields. Addressed all [PR comments](https://github.com/apache/iceberg/pull/14226)

Original PR got closed because of issues with branch.  Please let me know if there is a better way to handle this
cc @anuragmantri  @nastra and @amogh-jahagirdar

nastra commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -111,23 +125,20 @@ protected ActionsProvider actions() {
   @BeforeEach
   public void setupTableLocation() {
     this.tableLocation = tableDir.toFile().toURI().toString();
-    this.table = createATableWith2Snapshots(tableLocation);
```
I feel like we should set up the table here instead of in every single test method

gaborkaszab replied:
+1 If we extend the "bottom" `createTableWithSnapshots` at line 149 with the `format-version` property, then it will be valid for all the tests within this suite.

nastra commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -423,8 +442,10 @@ public void testPositionDeletes() throws Exception {
-  public void testPositionDeleteWithRow() throws Exception {
+  @TestTemplate
+  public void testV2PositionDeleteFilesWithRow() throws Exception {
+    assumeThat(formatVersion).isEqualTo(2);
```
same as above, you need to pass the `formatVersion` to FileHelpers.writeDeleteFile(..) and let the test run with all format versions

gaborkaszab commented:
I'd like to echo an observation I write to the [other PR](https://github.com/apache/iceberg/pull/14387#discussion_r2452531286): the runtime of this test suite seems long, introducing test dimensions could make it worse. I'd be careful here and only cover different format version where it matters and in fact adds to the test coverage.

Another thing: @adawrapub I saw that some of the tests aren't run with the new params (like testTableWithManyPartitionStatisticFile). If I'm not mistaken these are the ones that create the test table not by the @BeforeEach functionality but by calling `createMetastoreTable`.
 - Just noting, that these tests won't pick up the format-version param. 
 - Another comment that when I ran them parameterized I found that they aren't idempotent: the run for the first param succeeded, but the runs after failed. Probably some cleanup after the first run also cleaned up something needed by the runs after, because when I changed the table name to be unique between runs (e.g. adding the test param value to the name) then all the runs succeeded.

adawrapub commented:
Thank you for the feedback @nastra and @gaborkaszab. I have updated the PR based on your comments. 
Regarding the runtime of the test case, currently TestRewriteTablePathsAction is taking around 2 min to complete on my machine. But if it seems higher, we can split this test case into multiple classes, one class per format version. Please let me know what you think.

adawrapub commented:
> @adawrapub I'm a little confused on why you don't have failing tests with v3. Did you apply the diff I mentioned fully?

Hello @nastra Thank you for your response and patch. I did apply all your changes which caused some of the test cases to fail. My understanding is  Format v3+ uses PUFFIN format for position deletes. The positionDeletesReader method doesn't support PUFFIN format. Hence in test cases like testPositionDeleteWithRow, testPositionDeletes, testDeleteFrom,  I added assumeThat(formatVersion) to not run for v3 and v4. Let me know if there is a better way to fix this. I did update FileHelpers.writePosDeleteFile and FileHelpers.writeDeleteFile to take format version.

adawrapub commented:
> > Hello @nastra Thank you for your response and patch. I did apply all your changes which caused some of the test cases to fail. My understanding is Format v3+ uses PUFFIN format for position deletes. The positionDeletesReader method doesn't support PUFFIN format. Hence in test cases like testPositionDeleteWithRow, testPositionDeletes, testDeleteFrom, I added assumeThat(formatVersion) to not run for v3 and v4. Let me know if there is a better way to fix this. I did update FileHelpers.writePosDeleteFile and FileHelpers.writeDeleteFile to take format version.
> 
> Those remaining failing tests is the indicator of the underlying issue that needs to be fixed as part of this PR. We don't want to skip those tests for v3+ but we need to make rewriting table paths work for v3+, which is the scope of #13671 and which I also mentioned in [#14226 (comment)](https://github.com/apache/iceberg/pull/14226#issuecomment-3354917294). You need to apply a similar approach to what has been done in #11657 in order to allow reading v3 deletes and thus make those tests pass

@nastra As I understood DV files (Puffin format) store the referenced data file path in two separate locations:
Manifest Metadata: DeleteFile.referencedDataFile() field
Puffin Blob Metadata: "referenced-data-file" property inside the blob
The original implementation only updated the manifest metadata. The Puffin blob metadata still contained the old path, causing the DV reader to fail when applying deletes at the new location.

I have implemented a two-pronged update strategy. Please let me know if there is a better way

1. Manifest Metadata Update
Added ContentFileUtil.replaceReferencedDataFile() utility method
Created RewriteTablePathUtil.newPositionDeleteEntry() helper method
Updates the referencedDataFile field in manifest entries
2. Puffin Content Update
Implemented RewriteTablePathUtil.rewriteDVFile() method
Reads Puffin files, updates blob metadata properties, writes new files
Preserves the bitmap data while updating path references

nastra commented on core/src/main/java/org/apache/iceberg/util/ContentFileUtil.java:
```diff
@@ -130,6 +130,31 @@ public static Metrics replacePathBounds(
+   * @param targetPrefix target prefix that will replace it
+   * @return updated referenced data file path, or null if not applicable
+   */
+  public static String replaceReferencedDataFile(
```
probably best to keep this internally in `RewriteTablePathUtil`

adawrapub replied:
@nastra Updated. Please let know if it looks good to you

nastra commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -524,6 +516,55 @@ private static DeleteFile newEqualityDeleteEntry(
+   * @param targetPrefix target prefix that will replace it
+   * @return updated referenced data file path, or null if not applicable
+   */
+  private static String replaceReferencedDataFile(
```
```suggestion
  private static String rewriteReferencedDataFilePathForDV(
```

nastra commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -592,6 +641,66 @@ record = recordIt.next();
+                  blobMetadata.snapshotId(),
+                  blobMetadata.sequenceNumber(),
+                  blobData,
+                  null, // compression codec (keep uncompressed)
```
```suggestion
                  PuffinCompressionCodec.forName(blobMetadata.compressionCodec()),
```

nastra commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -592,6 +641,66 @@ record = recordIt.next();
+      throws IOException {
+    InputFile sourceFile = io.newInputFile(deleteFile.location());
+
+    try (org.apache.iceberg.puffin.PuffinReader reader =
```
can you please apply the below diff?

```
private static void rewriteDVFile(
      DeleteFile deleteFile,
      OutputFile outputFile,
      FileIO io,
      String sourcePrefix,
      String targetPrefix)
      throws IOException {
    List<Blob> rewrittenBlobs = Lists.newArrayList();
    try (PuffinReader reader = Puffin.read(io.newInputFile(deleteFile.location())).build()) {
      // Read all blobs and rewrite them with updated referenced data file paths
      for (Pair<org.apache.iceberg.puffin.BlobMetadata, ByteBuffer> blobPair :
          reader.readAll(reader.fileMetadata().blobs())) {
        org.apache.iceberg.puffin.BlobMetadata blobMetadata = blobPair.first();
        ByteBuffer blobData = blobPair.second();

        // Get the original properties and update the referenced data file path
        Map<String, String> properties = Maps.newHashMap(blobMetadata.properties());
        String referencedDataFile = properties.get("referenced-data-file");
        if (referencedDataFile != null && referencedDataFile.startsWith(sourcePrefix)) {
          String newReferencedDataFile = newPath(referencedDataFile, sourcePrefix, targetPrefix);
          properties.put("referenced-data-file", newReferencedDataFile);
        }

        // Create a new blob with updated properties
        rewrittenBlobs.add(
            new Blob(
                blobMetadata.type(),
                blobMetadata.inputFields(),
                blobMetadata.snapshotId(),
                blobMetadata.sequenceNumber(),
                blobData,
                PuffinCompressionCodec.forName(blobMetadata.compressionCodec()),
                properties));
      }
    }

    try (PuffinWriter writer =
        Puffin.write(outputFile).createdBy(IcebergBuild.fullVersion()).build()) {
      rewrittenBlobs.forEach(writer::write);
    }
  }
```

nastra commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -386,22 +408,23 @@ public void testDeleteDataFile() throws Exception {
   }
 
   private void runPositionDeletesTest(String fileFormat) throws Exception {
+    assumeThat(formatVersion).isEqualTo(2);
```
we shouldn't be skipping this for v3. Instead we need to pass the formatVersion further below so that the right type of delete file is created

nastra reviewed (approved):
LGTM, thanks for fixing this @adawrapub
